### PR TITLE
FxA webchannel support and account manager integration

### LINKED ISSUE #4238: AccountObserver: differentiate between sign-in and sign-up
We'll need this for telemetry purposes.

Also: currently there's a flag - `newAccount` - which differentiates between "sign-in/sign-up and restore". That's a potentially misleading name, in retrospect. For example, due to a mistake Fenix treats it as a flag differentiating between "sign-in and sign-up". Let's rename that to `isRestore`, or similar.

### PR DESCRIPTION
**Pull request by grigoryk**
## Part 1:
Preparatory work for allowing users to choose what to sync. We are about to
tie SyncConfig with FxA's "capabilities" webchannel API, which takes a set of
"engines" as strings. To simplify our integration work, this patch replaces an
untyped set of "stores" with a typed set of "engines" at the API level.

## Part 2:
(should have been split up further for easier reviewing, but things got a bit tangled)
This patch does a few things:
- adds a webextension for communicating with fxa web content
- provides a backend for said webextension, which understands webchannel communication protocol
and knows how to relay messages to the account manager, and query its state
- expands account manager a bit to support the webchannel authentication protocol
- adds a (rather detailed!) AuthType parameter to AccountObserver:onAuthenticated, instead of the very vague 'newAccount' flag
  - Closes #4238 

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by grigoryk** [comment]
Fenix side of this: https://github.com/mozilla-mobile/fenix/pull/4931 (note that fxa functionality is currently running on a dev box 2).

File: components/feature/accounts/src/main/java/mozilla/components/feature/accounts/WebChannelFeature.kt [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+        internal var ports = WeakHashMap<EngineSession, Port>()
+
+        /**
+         * Installs the WebChannel web extension in the provided engine.
+         *
+         * @param engine a reference to the application's browser engine.
+         */
+        fun install(engine: Engine) {
+            engine.installWebExtension(WEB_CHANNEL_EXTENSION_ID, WEB_CHANNEL_EXTENSION_URL,
```
Comment (csadilek) [inline components/feature/accounts/src/main/java/mozilla/components/feature/accounts/WebChannelFeature.kt]: This had to bring in a lot of code from the reader view feature. I've filed https://github.com/mozilla-mobile/android-components/issues/4297 to address the code duplication and come up with an abstraction.

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/fxawebchannel.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+/*
+Handle messages from native application, dispatch them to FxA via an event.
+*/
+port.onMessage.addListener((event) => {
+  window.dispatchEvent(new CustomEvent('WebChannelMessageToContent', {
+    detail: JSON.stringify(event)
+  }));
+});
+
+window.addEventListener("unload", (event) => { port.disconnect() }, false);
```
[the remark this answers is not in the document]
Comment (csadilek) [reply components/feature/accounts/src/main/assets/extensions/fxawebchannel/fxawebchannel.js]: We could just add comment here to take care of this later as part of https://github.com/mozilla-mobile/android-components/issues/2984

File: components/feature/accounts/src/main/java/mozilla/components/feature/accounts/WebChannelFeature.kt [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+ * This feature uses a web extension to communicate with FxA Web Content.
+ *
+ * @property context a reference to the context.
+ * @property customTabSessionId optional custom tab session ID, if feature is being used with a custom tab.
+ * @property engine a reference to application's browser engine.
+ * @property sessionManager a reference to application's [SessionManager].
+ * @property accountManager a reference to application's [FxaAccountManager].
+ */
+@Suppress("TooManyFunctions")
+class WebChannelFeature(
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/feature/accounts/src/main/java/mozilla/components/feature/accounts/WebChannelFeature.kt]: Agree!

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/fxawebchannel.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (grigoryk) [reply components/feature/accounts/src/main/assets/extensions/fxawebchannel/fxawebchannel.js]: Left a comment linking back to the issue.

**Comment by grigoryk** [comment]
@csadilek addressed your feedback.

**Comment by grigoryk** [comment]
reference-browser PR: https://github.com/mozilla-mobile/reference-browser/pull/895

**Comment by grigoryk** [comment]
bors r=csadilek

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/manifest.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
+  "manifest_version": 2,
+  "name": "Mozilla Android Components - Firefox Accounts WebChannel",
+  "version": "1.0",
+  "content_scripts": [
+    {
+      "matches": ["https://acounts.firefox.com/*"],
```
Comment (vladikoff) [inline components/feature/accounts/src/main/assets/extensions/fxawebchannel/manifest.json]: this is missing lcip domains, we should probably include them still

File: components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+     * [fxa-web-content] <--js events--> [fxawebchannel.js webextension] <--port messages--> [FxaWebChannelFeature]
+     *
+     * Overall message flow, as implemented by this class, is documented below. For detailed message descriptions, see:
+     * https://github.com/mozilla/fxa/blob/master/packages/fxa-content-server/docs/relier-communication-protocols/fx-webchannel.md
+     *
+     * [fxa-web-channel]            [FxaWebChannelFeature]         Notes:
+     *     loaded           ------>          |                  fxa web content loaded
+     *     fxa-status       ------>          |                  web content requests account status & device capabilities
+     *        |             <------ fxa-status-response         this class responds, based on state of [accountManager]
+     *     can-link-account ------>          |                  user submitted credentials, web content verifying if account linking is allowed
```
Comment (vladikoff) [inline components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt]: can link account should ago away?

File: components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+ * Boilerplate around installing and communicating with the extension will be cleaned up as part https://github.com/mozilla-mobile/android-components/issues/4297.
+ *
+ * @property context a reference to the context.
+ * @property customTabSessionId optional custom tab session ID, if feature is being used with a custom tab.
+ * @property engine a reference to application's browser engine.
+ * @property sessionManager a reference to application's [SessionManager].
+ * @property accountManager a reference to application's [FxaAccountManager].
+ */
+@Suppress("TooManyFunctions")
+class FxaWebChannelFeature(
```
Comment (vladikoff) [inline components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt]: the other class is called `Firefox Accounts Auth Feature` , this one just says `Fxa....`, does that matter?

**Pull request merged**